### ISSUE TITLE
Soft wrap happening at 500th character with soft wrap disabled

### ISSUE THREAD

**Issue by MiquelAdell** [atom/atom#13993]
### Prerequisites

* [x] Put an X between the brackets on this line if you have done all of the following:
    * Reproduced the problem in Safe Mode: http://flight-manual.atom.io/hacking-atom/sections/debugging/#using-safe-mode
    * Followed all applicable steps in the debugging guide: http://flight-manual.atom.io/hacking-atom/sections/debugging/
    * Checked the FAQs on the message board for common solutions: https://discuss.atom.io/c/faq
    * Checked that your issue isn't already filed: https://github.com/issues?utf8=✓&q=is%3Aissue+user%3Aatom
    * Checked that there is not already an Atom package that provides the described functionality: https://atom.io/packages

### Description

When soft-wrap is disabled I get line wraps at all lines at the 500th character.

### Steps to Reproduce

1. create a file with:
`012345678 012345678 012345678 012345678 012345678 012345678 012345678 012345678 012345678 012345678 012345678 012345678 012345678 012345678 012345678 012345678 012345678 012345678 012345678 012345678 012345678 012345678 012345678 012345678 012345678 012345678 012345678 012345678 012345678 012345678 012345678 012345678 012345678 012345678 012345678 012345678 012345678 012345678 012345678 012345678 012345678 012345678 012345678 012345678 012345678 012345678 012345678 012345678 012345678 012345678 012345678 `
(or open any file with long lines)

2. Disable soft wrap

**Expected behavior:**

The file should be displayed in one line

**Actual behavior:** 

The file is displayed in 2 lines

**Reproduces how often:** [What percentage of the time does it reproduce?]

Always

### Versions

You can get this information from copy and pasting the output of `atom --version` and `apm --version` from the command line. Also, please include the OS and what version of the OS you're running.

```
Atom    : 1.15.0
Electron: 1.3.13
Chrome  : 52.0.2743.82
Node    : 6.5.0
OS:  Mac OS Sierra 10.12.3 (16D32)
```

### Additional Information

If tried that on safe mode, with saved fles, with unsaved text pasted on a new window everything.
I've searched and found nothing about it so I guess it's not happening to everybody but I can't figure how this is happening to me.

**Comment by 50Wliu** [comment]
This is now intentional so that files with long lines do not freeze Atom.  See #13820 for more info.

**Comment by MiquelAdell** [comment]
Damm, thanks.
It's inconvenient right now for me (opening that did not hang before, not minifed files but with some 600~ text lines) but I guess I'll curse a lot less when opening a minified one by mistake.

Thanks.

**Comment by d3rrial** [comment]
Can this also be deactivated? It's annoying AF in the scenarios which I use Atom for.

**Comment by Zireael07** [comment]
@d3rrial: No it can't, see @50Wliu's explanation above.

**Comment by garretwilson** [comment]
I just ran into this on 1.15.0 ia32. This is horrible. If I turn off line wrapping, I expect lines not to be wrapped.

This should be reopened. Basically you're saying, "oops, we ran into some sort of limitation, so we'll arbitrarily wrap your lines even though you've explicitly told us not to". (And this "limitation" seems to be around 500 characters. I could understand if we were running on a Commodore 64, but seriously...)

I can't control line-wrapping on my editor? Really?

I'm not so irritated that you ran into a bug---that happens. I'm amazed and dumbfounded that you would close this issue and essentially say, "you'll have to live with it."

**Comment by allencrawford** [comment]
Totally agree.  As a compromise it seems that you could leave the original behavior when the file size is really small.  I've got a one line regex in a config file that's over 700 characters long, but the file is maybe 10 lines long total.  It is irritating when it soft wraps on a space and I have to wonder if the space is really in my line or not in an important config file where spaces matter.

**Comment by 50Wliu** [comment]
Unfortunately, that won't work.  Atom will hang on files with very long lines, regardless of file size.  A lot of reports were with minified jQuery, which is only around 3kb (I think).

**Comment by Zireael07** [comment]
@50Wliu: Can you somehow detect how long the line is, and don't apply the auto-wrap if the length is less than some critical value?

**Comment by 50Wliu** [comment]
@Zireal07 Maybe I'm misunderstanding your question, but isn't that what Atom is already doing with 500 characters being the critical value?

**Comment by d3rrial** [comment]
@50Wliu 500 is an arbitrarily chosen value that they chose because most singular lines of code (apparently) do not exceed that and it was having stable results in their testing. They could've also done with 2000 or 3000 characters, but variance between systems may or may not compromise stability at those cases. I'm assuming that their thought is better have some **extremely** annoying line-wrapping than having someone lose unsaved files because they opened a file that was too long and crashed the editor.

**Comment by garretwilson** [comment]
To summarize @d3rrial: "Atom: We're the editor that makes you choose between losing data or being extremely annoyed." 😞

**Comment by Zireael07** [comment]
@50Wliu: I am thinking "don't wrap at 500 if the line is shorter than, I dunno, 700"? 

A note on the soft wrap disable option informing users of the auto-wrap would be a good idea, too.

**Comment by 50Wliu** [comment]
> A note on the soft wrap disable option informing users of the auto-wrap would be a good idea, too.

I think we'd be open for a PR for this. @lee-dohm

**Comment by lee-dohm** [comment]
We don't feel that having an option that all of a sudden makes Atom unstable is a valid choice to give people, no matter how many warnings that are placed around it. It is proven by research time and again that people don't read the explanatory text around things until something goes wrong. By then it is too late for the options being discussed here, as I understand them.

> and essentially say, "you'll have to live with it."

If you don't like the limitation or the value that we have chosen, you have the option of compiling Atom from source with the value set to whatever you wish it to be set to.

**Comment by d3rrial** [comment]
@lee-dohm I believe that @Zireael07 meant something along the lines of:

"Disabling Line-Wrapping doesn't actually disable Line-Wrapping. Re-Compile from source if you want that."

as the warning.

**Comment by allencrawford** [comment]
There are many, many things to like about Atom, but the layperson such as myself just wonders why Notepad++ can handle long lines and large files so well but Atom can't.  I'll keep the faith that you'll get there, but with all the configuration options that exist I'm not sure why this one can't be configured (without a recompile).

**Comment by d3rrial** [comment]
@allencrawford Notepad++ isn't based on node.js and V8. It's totally reasonable that it can't perform the same as a native application.

**Comment by mscottford** [comment]
I just bumped up against this, but I've noticed that when I'm on a line that wraps at the 500 character boundary there's no ability to "select to the end of the line" all of the options I've tried only "select to the end of the virtual line". `^-E` correctly jumps to the end of the actual line but `^-SHIFT-E` only selects to the end. Is that intended? I'm often editing long lines like this, and I'm used to using `CMD-D` to select a bunch of them and then reformat them all the same way. Before the 500 char boundary things were slow but manageable. I'm happy to open a separate issue if that's what makes the most sense.

**Comment by toferj** [comment]
This is annoying as hell. Sublime Text does not have this problem... so why does Atom? Why can't it be fixed? I have some lines that contain Looooooong comma delimited lists, and it annoys me to no end to see these wrap. Is there any plan to work on making this behaviour go away, or have the developers just given up and settled on this really unsatisfying compromise?

**Comment by 50Wliu** [comment]
> Why can't it be fixed?

The great thing about open source is that everyone can contribute to the codebase!  If you have ideas for how to improve Atom's performance when rendering long lines, then we'd love if you could open a new topic on [Discuss](https://discuss.atom.io) to garner feedback or a PR if you have a potential code solution.

**Comment by vpal** [comment]
This is an extremely sad story...

**Comment by toferj** [comment]
Sad indeed... I was hoping that someone could tell me *why* Atom chokes on long lines. Why is there a performance problem.  But all the open source community seems to wants to point out is that anybody can contribute to the code base. You want to know why? Look it up for yourself. 

"Atom: the hackable editor"... problem is I don't have the time to waste on it, or the desire. I just want the editor I use to work.  I've gone back to Sublime Text. ~$70 is super-cheap for a license, and it performs much better. 

"Atom: I don't see what all the fuss is about"

**Comment by stankay** [comment]
At least could the this feature configurable with note "Use it at your own risk"

**Comment by Ben3eeE** [comment]
It is configurable https://github.com/atom/atom/pull/14495

**Comment by garretwilson** [comment]
> It is configurable #14495

It's a little confusing poring through closed issues to find out how to turn off behavior we never asked to be turned on.

I looked in the configuration; I didn't see any option for this. (I did see that I have turned _off_ all line wrapping options available.)

So going over #14495, I'm still not clear; do I set the `maxScreenLineLength` in the config? Where do I put it --- in the `editor:` section? What do I set it to in order to turn off line wrapping altogether?

Thanks for clarifications.

**Comment by vpal** [comment]
Does 1.21.1 already support this feature?
I have
```
...
  editor:
    fontSize: 15
    invisibles: {}
    maxScreenLineLength: 5000
    preferredLineLength: 120
    showIndentGuide: true
    showInvisibles: true
...
```
in my config.cson and it has no effect.

**Comment by rsese** [comment]
The setting is not available in 1.21, it's in 1.22 currently in beta.  The setting is `Settings > Editor > Max Screen Line Length`.

**Comment by craigmiller160** [comment]
Wow... I'm going to jump on the bandwagon of hating on this app. I really wanted to like it, but I am also working on a project with a flat file format where line delimiters indicate new records. Wrapping the text makes the format MUCH harder to read. Stop making excuses, this is absolutely unacceptable from a product like this. If it's a limitation of Node.js, well guess what, the developers made the decision to use that as their platform, so they are responsible for all of its limitations. Abandoning Atom now and going back to Sublime. Bye.
